File: applications/notebook_harness/widget-src/src/player_list_widget.js
// Player roster panel shown beside the route graph. Renders one row per
// seat from the `players` trait ({id, name, color} dicts, the shape
// HarnessGame.roster() produces): a swatch in the seat color - the same
// color used for that player's claim markers on the board - plus a name.
function render({ model, el }) {
    el.classList.add("player-list-widget");

    const draw = () => {
        el.replaceChildren();

        const heading = document.createElement("div");
        heading.className = "player-list-heading";
        heading.textContent = "Players";
        el.appendChild(heading);

        (model.get("players") || []).forEach((player) => {
            const row = document.createElement("div");
            row.className = "player-list-row";

            const swatch = document.createElement("span");
            swatch.className = "player-list-swatch";
            swatch.style.background = player.color || "#999999";

            const name = document.createElement("span");
            name.className = "player-list-name";
            name.textContent = player.name || player.id;

            row.appendChild(swatch);
            row.appendChild(name);
            el.appendChild(row);
        });
    };

    draw();
    model.on("change:players", draw);
}

export default { render };

File: applications/notebook_harness/player_list_widget.py
from __future__ import annotations

from importlib.resources import files

import anywidget
import traitlets


class PlayerListWidget(anywidget.AnyWidget):
    """Roster panel shown beside the board: one row per seat with the seat's
    color swatch (matching that player's claim markers on the map) and name.

    ``players`` takes the list of ``{id, name, color}`` dicts produced by
    ``HarnessGame.roster()``. See ``applications/notebook_harness/widget-src/``
    for the JS source this is bundled from.
    """

    _esm = files("notebook_harness").joinpath("static/player_list_widget.js")
    _css = files("notebook_harness").joinpath("static/player_list_widget.css")

    players = traitlets.List([]).tag(sync=True)
